spec/integration/rack_app/middleware_spec.rb

```diff
@@ -235,4 +235,424 @@ def handle(req, res)
+
+                [status, headers, body]
+              end
+            end
+          end
+        end
+      RUBY
+
+      write "lib/test_app/middleware/scope_identifier.rb", <<~RUBY
+        module TestApp
+          module Middleware
+            class ScopeIdentifier
+              def self.inspect
+                "<Middleware::API::ScopeIdentifier>"
+              end
+
+              def initialize(app, scope)
+                @app = app
+                @scope = scope
+              end
+
+              def call(env)
+                status, header, body = @app.call(env)
+                header["X-Identifier-" + @scope] = "true"
+                [status, header, body]
+              end
+
+              def inspect
+                "Scope identifier: " + @scope.inspect
+              end
+            end
+          end
+        end
+      RUBY
+
+      write "config/routes.rb", <<~RUBY
+        require "test_app/middleware/elapsed"
+        require "test_app/middleware/authentication"
+        require "test_app/middleware/rate_limiter"
+        require "test_app/middleware/api_version"
+        require "test_app/middleware/api_deprecation"
+        require "test_app/middleware/scope_identifier"
+
+        module TestApp
+          class Routes < Hanami::Routes
+            use TestApp::Middleware::Elapsed
+            use TestApp::Middleware::ScopeIdentifier, "Root"
+            root to: ->(*) { [200, {"Content-Length" => "4"}, ["Home (complex app)"]] }
+
+            mount ->(*) { [200, {"Content-Length" => "7"}, ["Mounted"]] }, at: "/mounted"
+
+            slice :admin, at: "/admin" do
+              use TestApp::Middleware::Authentication
+              use TestApp::Middleware::ScopeIdentifier, "Admin"
+
+              root to: "home.show"
+            end
+
+            # Without leading slash
+            # See: https://github.com/hanami/api/issues/8
+            scope "api" do
+              use TestApp::Middleware::RateLimiter
+              use TestApp::Middleware::ScopeIdentifier, "API"
+
+              root to: ->(*) { [200, {"Content-Length" => "3"}, ["API"]] }
+
+              # slice :api_v1, at: "/v1" do
+              #   use TestApp::Middleware::ApiVersion
+              #   use TestApp::Middleware::ApiDeprecation
+              #   # use TestApp::Middleware::ScopeIdentifier, "API V1"
+
+              #   root to: "home.show"
+              # end
+            end
+          end
+        end
+      RUBY
+
+      write "slices/admin/actions/home/show.rb", <<~RUBY
+        module Admin
+          module Actions
+            module Home
+              class Show < Hanami::Action
+                def handle(req, res)
+                  res.body = "Hello from admin (User ID " + req.env['AUTH_USER_ID'] + ")"
+                end
+              end
+            end
+          end
+        end
+      RUBY
+
+      # write "slices/api_v1/actions/home/show.rb", <<~RUBY
```

Comment:
@timriley If I uncomment this file, I get the following error:

```shell
⚡ bundle exec rspec ./spec/integration/rack_app/middleware_spec.rb:343
Failures:

  1) Hanami web app with complex app uses Rack middleware
     Failure/Error: config.public_send(name, *args, &block)

     Dry::Configurable::FrozenConfig:
       Cannot modify frozen config
# ...
```

I don't understand why the other action above doesn't cause this problem, but this one does.